### PR TITLE
Issue96 - stamping of generic classes fixed

### PR DESCRIPTION
**Pull request by mateusz-holenko**

### CODE REVIEW

File: Migrant/Helpers.cs [hunk 1]
```diff
@@ -236,7 +236,31 @@ internal static int GetSurrogateFactoryIdForType(Type type, InheritanceAwareList
             }
             return -1;
         }
-        
+
+        internal static bool IsOpenedGenericType(Type type)
```
Comment (konrad-kruczynski) [inline Migrant/Helpers.cs]: Hmm, maybe "open"? http://english.stackexchange.com/a/141552

File: Migrant/PrimitiveReader.cs [hunk 2]
```diff
@@ -151,7 +154,7 @@ public sbyte ReadSByte()
         public short ReadInt16()
         {
 #if DEBUG
-            if(PrimitiveWriter.DontUseIntegerCompression)
+            if(Serializer.DISABLE_VARINTS)
```
Comment (konrad-kruczynski) [inline Migrant/PrimitiveReader.cs]: Why such casing?

File: Tests/SerializationTests.cs [hunk 3]
```diff
@@ -1015,6 +1015,43 @@ public void ShouldSerializeClassWithGenericMethodDelegate()
...
+        public void ShouldSerializeClassWithArrayOfNestedGenericArguments()
+        {
+            var serializer = new Serializer(GetSettings());
+
+            var obj = new GenericClassWithNestedGenericArgument<string>("test");
+            using(var stream = new MemoryStream())
+            {
+                serializer.Serialize(obj, stream);
+                stream.Seek(0, SeekOrigin.Begin);
+                using (var file = File.OpenWrite("stream.dump"))
```
Comment (konrad-kruczynski) [inline Tests/SerializationTests.cs]: Leftover?

File: Tests/VersionToleranceTests.cs [hunk 4]
```diff
@@ -183,6 +187,60 @@ public void TearDown()
...
+            // 32 - ' ', 126 - '~'
+            var interestingBytes = array.Where(b => b >= 32 && b <= 126).ToArray();
+            return Encoding.ASCII.GetString(interestingBytes);
+        }
+
+        private static int NumberOfOccurences(string src, string pattern)
+        {
+            var n = 0;
+            var count = 0;
+            while ((n = src.IndexOf(pattern, n)) != -1)
```
Comment (konrad-kruczynski) [inline Tests/VersionToleranceTests.cs]: Additional space.

File: Tests/SerializationTests.cs (hunk 3 above)
Comment (mateusz-holenko) [inline Tests/SerializationTests.cs]: Indeed.

**Comment by mateusz-holenko** [comment]
I simplified the format, so that stamping array is done only when necessary - i.e., when we deal with a type containing generic parameter (directly, being an argument of generic types or element type of array). Please review my changes again.

File: Migrant/VersionTolerance/ArrayDescriptor.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+//
+// *******************************************************************
+using System;
+using System.Collections.Generic;
+
+namespace Antmicro.Migrant.VersionTolerance
+{
+    public class ArrayDescriptor
+    {
+        public static ArrayDescriptor EmptyRanks { get { return new ArrayDescriptor(typeof(void), new int[0]); } }
```
Comment (konrad-kruczynski) [inline Migrant/VersionTolerance/ArrayDescriptor.cs]: I think there should be only one instance of EmptyRanks. The field should be automatic `get; private set` and inintialized in a static ctor.

File: Migrant/VersionTolerance/ArrayDescriptor.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
+﻿// *******************************************************************
```
Comment (konrad-kruczynski) [inline Migrant/VersionTolerance/ArrayDescriptor.cs]: Why isn't the array descriptor read from the stream by itself?

Comment (mateusz-holenko) [inline Migrant/VersionTolerance/ArrayDescriptor.cs]: Unfortunately, `ArrayDescriptor` is constructed from two parameters: `elementType` and `ranks` array. The first one is read from stream much earlier, so it is not possible to read whole descriptor at once. As a result I would stick with current solution.

File: Migrant/VersionTolerance/ArrayDescriptor.cs (hunk 5 above)
Comment (mateusz-holenko) [inline Migrant/VersionTolerance/ArrayDescriptor.cs]: Aye aye

**Pull request merged**